### PR TITLE
UPSTREAM: <carry>: with shard prefix support 

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What this PR does / why we need it:

This PR adds support for setting a shard prefix to the storage layer.
The prefix is used by the cache server.
The storage layer has to understand both the cluster and the shard prefixes
because the database can be shared between the kcp and the cache server.

Once a shard name is set, the storage layer will change the key under which a resource is stored, for example:

amber/root:org/default/secret
amber/root:org/secret

Where:

- amber:     is a shard name
- root:org:  is a cluster name
- default:   is a namespace name
- secret:    is a name of a resource

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 1]
```diff
@@ -238,15 +238,23 @@ const (
 // to resource directories enforcing namespace rules.
 func NoNamespaceKeyRootFunc(ctx context.Context, prefix string) string {
 	key := prefix
+	shard, err := genericapirequest.ValidShardFrom(ctx)
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: we need a better way of plugging a shard name into the storage layer.
something that the kcp and the cache server could customize upon initialization.

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+type shardKey int
+
+const (
+	// shardKey is the context key for the request namespace.
+	shardContextKey shardKey = iota
+)
+
+type Shard struct {
+	// Name is the name of the cluster.
+	Name string
```
Comment (stevekuznetsov) [inline staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go]: Given that we (likely) have shard name requirements, and `*` will not be a valid shard name, is there any reason not just have `type Shard string`? Otherwise, what is the meaning of:

```go
Shard{
  Name: "*",
  Wildcard: false,
}
```

or

```go
Shard{
  Name: "sapphire",
  Wildcard: true,
}
```
?

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go]: a request like `'https://localhost:6443/shards/*/clusters/*/apis/apis.kcp.dev/v1alpha1/apiexports'` will set
```
Shard{ 
  Name: "", 
  Wildcard: true
 }
```

a request like `https://localhost:6443/shards/sapphire/clusters/system:sapphire/apis/apis.kcp.dev/v1alpha1/apiexports'` will set
```
Shard{ 
  Name: "sapphire", 
  Wildcard: false
 }
```

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go (hunk 1 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: initially, I wanted to plug in different implementations but I realised the kcp and the cache server might share the same database instance.

**Comment by p0lyn0mial** [comment]
/assign @stevekuznetsov @sttts 

ready for review, PTAL.

**Comment by p0lyn0mial** [comment]
proof PR https://github.com/kcp-dev/kcp/pull/1851

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go (hunk 2 above)
Comment (stevekuznetsov) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go]: Right, those are the happy paths. My question was about the others, which are (I think) not possible. In e.g. https://github.com/kcp-dev/kcp/pull/1841 you have a `New('*")` value for wildcard... seems like this could also ?

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go]: I do use `New(*)` on the client side just to set the right URL path, then on the server side, it gets mapped to `Wildcard: true, Name: ""`. 

I think we need both fields. An empty `Name:"", Wildcard: false` indicates no shard was specified.

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	shardContextKey shardKey = iota
+)
+
+// Shard holds data that describes a shard.
+type Shard struct {
+	// Name is the name of the shard.
+	Name string
+
+	// If true the query applies to all shards
+	Wildcard bool
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go]: why this complex struct again? Why not "empty shard string = wildcard" or "* shard string = wildcard" ?

Note: we should do the same for the cluster.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go [hunk 4]
```diff
@@ -54,14 +60,20 @@ func adjustClusterNameIfWildcard(cluster *genericapirequest.Cluster, crdRequest
...
+		}
 	} else {
-		regex = wildcardClusterNameRegex
+		regex = wildcardShardClusterNameRegex
 	}
 
 	matches := regex.FindStringSubmatch(keyWithoutPrefix)
-	if len(matches) >= 2 {
+	if len(matches) >= 3 {
+		return logicalcluster.New(matches[2])
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:77]: what does this mean? Some high-level comments would help

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go [hunk 5]
```diff
@@ -133,8 +134,12 @@ func (w *watcher) Watch(ctx context.Context, key string, rev int64, recursive bo
 	if err != nil {
 		return nil, err
 	}
+	shard, err := genericapirequest.ValidShardFrom(ctx)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go]: what does this return if there is no shard in the context?

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go (hunk 3 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go]: we need both, I can replace with a string but then we need to compare to `*`

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go (hunk 5 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go]:
```
// ValidShardFrom returns the value of the shard key from the context.
// An empty value is returned if there's no shard key,
// or if the shard name is empty, and it wasn't a wildcard request.
```

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go (hunk 4 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:77]: have a look at
```
	// matches shard/cluster/remainder, capturing shard and cluster.
	//
	// Example: amber/root:org:ws/some-namespace/some-name
	// Example: amber/root:org:ws/some-name
	wildcardShardClusterNameRegex = regexp.MustCompile(`^([^\/]+)\/([^\/]+)\/.+`)
```

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	if !ok {
+		return nil
+	}
+	return &shard
+}
+
+// ValidShardFrom returns the value of the shard key from the context.
+// An empty value is returned if there's no shard key,
+// or if the shard name is empty, and it wasn't a wildcard request.
+func ValidShardFrom(ctx context.Context) (*Shard, error) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go]: why "Valid"? ShardFromOrEmpty would be more idiomatic

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go]: true

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
+package request
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go:17]: do we have the boilerplate for our files?

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go [hunk 8]
```diff
@@ -33,6 +33,12 @@ var (
 	// Example: root:org:ws/some-name
 	wildcardClusterNameRegex = regexp.MustCompile(`^([^/]+)\/.+`)
 
+	// matches shard/cluster/remainder, capturing shard and cluster.
+	//
+	// Example: amber/root:org:ws/some-namespace/some-name
+	// Example: amber/root:org:ws/some-name
+	wildcardShardClusterNameRegex = regexp.MustCompile(`^([^\/]+)\/([^\/]+)\/.+`)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:40]: this does not match `amber/root:org:ws/some-name`, does it?

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:77]: hard to reason about. Why not move the matching and result logic into the upper if-else construct?

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go (hunk 7 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go:17]: oh, right, I think we could add k8s

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go (hunk 8 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:40]: I think it does, let me check.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:40]: looks okay - https://go.dev/play/p/OFjiSce2jCC

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go (hunk 7 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go:17]: `storage_patch` has `Copyright 2022 The KCP Authors.` Going to add kcp boilerplate

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go:17]: done

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go (hunk 4 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:77]: done.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go [hunk 9]
```diff
@@ -45,26 +51,31 @@ var (
...
 // it is a CR partial metadata request (prefix/identity/clusterName/remainder) or not (prefix/clusterName/remainder).
-func adjustClusterNameIfWildcard(cluster *genericapirequest.Cluster, crdRequest bool, keyPrefix, key string) logicalcluster.Name {
+func adjustClusterNameIfWildcard(shard genericapirequest.Shard, cluster *genericapirequest.Cluster, crdRequest bool, keyPrefix, key string) logicalcluster.Name {
 	if cluster.Name != logicalcluster.Wildcard {
 		return cluster.Name
 	}
 
 	keyWithoutPrefix := strings.TrimPrefix(key, keyPrefix)
 
 	var regex *regexp.Regexp
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go]: move this into the first if.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go [hunk 10]
```diff
@@ -45,26 +51,31 @@ var (
...
+		if len(matches) >= 2 {
+			return logicalcluster.New(matches[1])
+		}
+		return logicalcluster.Name{}
 	}
 
-	matches := regex.FindStringSubmatch(keyWithoutPrefix)
-	if len(matches) >= 2 {
-		return logicalcluster.New(matches[1])
+	matches := wildcardShardClusterNameRegex.FindStringSubmatch(keyWithoutPrefix)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:75]: Am still confused here. Why is this not `wildcardClusterNameRegex`?

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:75]: we cannot use `wildcardClusterNameRegex` because it won't capture a cluster name, only a shard name.
for example: given the key of `amber/root:org:ws/some-name` it will capture `amber` as a cluster name

basically, once this PR merges we will have two ways of storing data.
1. the kcp server that always adds a cluster name 
2. the cache-server that always adds a shard name and a cluster name.

when reading data back we need to take those two modes into account.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:75]: that means the cache-server won't read data stored by the kcp server and vice versa.

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go (hunk 2 above)
Comment (stevekuznetsov) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go]: How is `Name:"", Wildcard: false` different from `New("")`?

Comment (stevekuznetsov) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go]: My argument: the `Wildcard` field as you have written it is logically equivalent in every case to `obj.Name == "*"`

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go (hunk 10 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:75]: the above only applies to shard = "*", when a shard is specified then we change the prefix (https://github.com/kcp-dev/kubernetes/pull/96/files#diff-e52c66f0ffc19a41478dc44cd39fc05fe64f371d73f9db362c9df6047b94b810R239) and if the cluster was not specified will use `wildcardClusterNameRegex`

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:75]: and finally if a shard and a cluster were specified we will simply
```
f cluster.Name != logicalcluster.Wildcard {
		return cluster.Name
}
```

I want to capture all those details in https://github.com/kcp-dev/kcp/issues/1866

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go (hunk 2 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go]: @stevekuznetsov I updated the type to `type Shard string`, now `Shard("*")` indicates wildcard and `Shard("")` indicates a request without a shard name specified.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go]: I hope it is less confusing now.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go (hunk 9 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go]: done.

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+// WithShard returns a context that holds the given shard.
+func WithShard(parent context.Context, shard Shard) context.Context {
+	return context.WithValue(parent, shardContextKey, shard)
+}
+
+// ShardFrom returns the value of the shard key in the context, or an empty value if there is no shard key.
+func ShardFrom(ctx context.Context) Shard {
+	shard, ok := ctx.Value(shardContextKey).(Shard)
```
Comment (stevekuznetsov) [inline staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go:56]: nit: this simplifies to

```go
func ShardFrom(ctx context.Context) Shard {
	return ctx.Value(shardContextKey).(Shard)
}
```

[playground](https://go.dev/play/p/1HEbThSNRjg)

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 12]
```diff
@@ -238,15 +238,24 @@ const (
...
+	shard := genericapirequest.ShardFrom(ctx)
+	if shard.Wildcard() {
+		return key
+	}
 	cluster, err := genericapirequest.ValidClusterFrom(ctx)
 	if err != nil {
 		klog.Errorf("invalid context cluster value: %v", err)
 		return key
 	}
-	if cluster.Wildcard {
```
Comment (stevekuznetsov) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:246]: This is a bit hard to follow. Is it equivalent:

```go
if !shard.Empty() {
    key += "/" + shard.String()
}

if !cluster.Wildcard {
    key += "/" + cluster.Name.String()
}

return key
```

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go [hunk 13]
```diff
@@ -45,26 +51,31 @@ var (
 // adjustClusterNameIfWildcard determines the logical cluster name. If this is not a wildcard list/watch request,
 // the cluster name is returned unmodified. Otherwise, the cluster name is extracted from the key based on whether
 // it is a CR partial metadata request (prefix/identity/clusterName/remainder) or not (prefix/clusterName/remainder).
-func adjustClusterNameIfWildcard(cluster *genericapirequest.Cluster, crdRequest bool, keyPrefix, key string) logicalcluster.Name {
+func adjustClusterNameIfWildcard(shard genericapirequest.Shard, cluster *genericapirequest.Cluster, crdRequest bool, keyPrefix, key string) logicalcluster.Name {
```
Comment (stevekuznetsov) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:54]: IMHO this doesn't really make sense to do with regex. What about something like:


```go
 func adjustClusterNameIfWildcard(shard genericapirequest.Shard, cluster *genericapirequest.Cluster, crdRequest bool, keyPrefix, key string) logicalcluster.Name {
    if cluster.Name != logicalcluster.Wildcard || !cluster.Wildcard { // TODO: fix this duplicity, as well
        return cluster.Name
    }
    
    keyWithoutPrefix := strings.TrimPrefix(key, keyPrefix)
    
    var extractCluster func([]string) logicalcluster.Name
    extractor := func(numParts, clusterPart int) func([]string) logicalcluster.Name {
        return func([]string) logicalcluster.Name {
            if len(parts) < numParts {
                klog.Warnf("shard=%s cluster=%s invalid key=%s had %d parts, not %d", shard, cluster, keyWIthoutPrefix, len(parts), numParts)
                return logicalcluster.Name{}
            }
            return logicalcluster.New(parts[clusterPart])
        }
    }
    
    switch {
    case shard.Wildcard():
        // expecting shardName/clusterName/<remainder>
        extractCluster = extractor(3,1)
    case cluster.PartialMetadataRequest && crdRequest:
        // expecting 2699f4d273d342adccdc8a32663408226ecf66de7d191113ed3d4dc9bccec2f2/root:org:ws/<remainder>
        // OR customresources/root:org:ws/<remainder>
        extractCluster = extractor(3,1)
    case cluster.Wildcard():
        // expecting root:org:ws/<remainder>
        extractCluster = extractor(2,0)
    default:
        panic("bad programmer")
    }
    
    parts := strings.Split(keyWithoutPrefix, "/")
    return extractCluster(parts)
}
```

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go (hunk 11 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go:56]: it could, it will just panic if the value under the key cannot be cast to `Shard`

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go (hunk 13 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:54]: uh, maybe :) It seems like what we have right now requires less code. 
Maybe I could revisit this in the future and try to rewrite (refactor) it.

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go (hunk 12 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:246]: it seems it is :)

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:246]: done

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go (hunk 11 above)
Comment (stevekuznetsov) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go:56]: The `shardContextKey` is private to this package. The only way that it would panic is if someone wrote code _in this package_ to add the wrong thing. /shrug

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go (hunk 13 above)
Comment (stevekuznetsov) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:54]: Less code how exactly? Half of what I wrote were comments ;) Existing approach with your diff is L29-L80=51LoC. Mine is 35LoC

I think using regex to `strings.Split(key, "/")` makes the current approach really hard to follow, especially as we grow more cases.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:54]: Is it about taste or a big improvement? Either way is not particularly pretty. But this func does not decide about maintainability of the code, does it? Would be fine with a minimal change. Would pretty much prefer a unit test for either implementation ;-)

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:54]: @stevekuznetsov would it be okay to add unit tests and leave using the regexps for now?

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go (hunk 11 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go:56]: yes, I just don't like when the code panics but sure I can change it.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go (hunk 13 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:54]: btw: do we actually run k8s unit tests :) ?

Comment (stevekuznetsov) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:54]: I don't think it's taste, no, because using regex to split a string on `/` is really hard to understand - I think we all tripped over understanding this function when reviewing it, no? Indexing into a slice seems ... slightly more straightforward, as well as using `switch/case` to be exhaustive on options?

No, we do not yet run the tests.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go (hunk 10 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:75]: ack. So the old case is in the if above.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go (hunk 13 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:54]: @p0lyn0mial can you create an issue to track Steve's suggestion above? Would not like to block this PR on a refactoring that is kind of independent of this change because pre-existing.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:54]: keep the suggested code snipped, or even create the follow-up PR just after if it compiles.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch_test.go [hunk 14]
```diff
@@ -67,7 +67,66 @@ func TestAdjustClusterNameIfWildcard(t *testing.T) {
...
+		})
+	}
+}
+
+func TestAdjustClusterNameIfWildcardWithShardSupport(t *testing.T) {
+	tests := map[string]struct {
+		clusterWildcard bool
+		shardWildcard   bool
+		prefix          string
+		builtInType     bool
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch_test.go]: can this be pure table test with inputs and outputs? Irritating there is no "expected*" in here. Just put `genericapirequest.Cluster` and `genericapirequest.Shard` in here, no need for abstraction.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go (hunk 13 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go:54]: done https://github.com/kcp-dev/kcp/issues/1875

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch_test.go (hunk 14 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch_test.go]: updated

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go (hunk 11 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go:56]: and it broke the kcp server
```
I0901 10:30:42.331404   38583 etcd3.go:306] Start monitoring storage db size metric for endpoint https://localhost:45743 with polling interval 30s
        panic: interface conversion: interface {} is nil, not request.Shard
        
        goroutine 1 [running]:
        k8s.io/apiserver/pkg/endpoints/request.ShardFrom(...)
        	/go/pkg/mod/github.com/p0lyn0mial/kubernetes/staging/src/k8s.io/apiserver@v0.0.0-20220901092652-1cf194952b52/pkg/endpoints/request/context_shard.go:56
 ```

**Review by sttts — APPROVED** [review]
